File: 2024/Day2/red_nosed_reports/src/main.rs
mod handlers;

use std::{fs, io::{self, BufRead}};
use handlers::red_nosed_report_context::RedNosedReportContext;

fn main() {
    siphon_safe_reports();
}

fn siphon_safe_reports(){
    println!("Collecting safe reports...");

    // Open the unusual data file
    match fs::File::open("unusual_data.txt"){
        Ok(file) => {

            // Define buffer to start reading each line of the file
            let reader = io::BufReader::new(file);

            let mut safe_report_count = 0;

            // Loop through each report to analyze
            for line in reader.lines(){

                // Check if the report is safe
                match line {
                    Ok(report) => {

                        let levels: Vec<i32> = report
                            .split_whitespace()
                            .map(|s| s
                                .parse()
                                .unwrap())
                            .collect();

                        let increasing_levels = levels.windows(2).all(|level| 
                            level[0] < level[1] && (level[1] - level[0]) <= 3);

                        let decreasing_levels = levels.windows(2).all(|level|
                            level[0] > level[1] && (level[0] - level[1]) <= 3);

                        if increasing_levels || decreasing_levels{
                            safe_report_count += 1;
                            continue;
                        }

                        let report_length = levels.len();
                        let mut level_position = 0;

                        while level_position < report_length{

                            let mut unsafe_levels = levels.clone();
                            unsafe_levels.remove(level_position);

                            let is_increasing = unsafe_levels.windows(2).all(|level|
                                level[0] < level[1] && (level[1] - level[0]) <= 3);

                            
                            let is_decreasing = unsafe_levels.windows(2).all(|level|
                                level[0] > level[1] && (level[0] - level[1]) <= 3);

                            if is_increasing || is_decreasing {
                                safe_report_count += 1;
                                break;
                            }

                            level_position += 1;

                        }
                    }
                    Err(error) => {
                        println!("Error occurred collecting safe reports ({error})")
                    }
                }
            }

            println!("{}", safe_report_count);

        } Err(error) => {
            
        }
    }
}

File: 2024/Day2/Python/red_nosed_reports/main.py
from report import Report

def main():
    reports = collect_reports()
    fix_problem_dampener(reports)

    print(len([report for report in reports if report.is_safe == True]))

def collect_reports() -> list[Report]:

    reports = []

    with open("unusual_data.txt", "r") as file:
        for report in file:
            levels = list(map(int, report.strip().split()))

            max_difference = 3

            is_increasing = is_level_increasing(levels, max_difference)
            
            is_decreasing = is_level_decreasing(levels, max_difference)
            
            if (is_increasing or is_decreasing):
                reports.append(Report(list(map(int, levels)), True))
            else:
                reports.append(Report(list(map(int, levels)), False))

    return reports

def fix_problem_dampener(reports: list[Report]):
    
    max_difference = 3

    for report in [report for report in reports if report.is_safe == False]:

        for level_index in range(len(report.levels)):
            levels_copy = report.levels.copy()

            del levels_copy[level_index]

            for level in range(len(levels_copy)):

                is_increasing = is_level_increasing(levels_copy, max_difference)

                is_decreasing = is_level_decreasing(levels_copy, max_difference)

                if (is_increasing or is_decreasing):
                    report.is_safe = True
            
def is_level_increasing(levels: list[int], max_difference: int) -> bool:
    return all(level < next_level and next_level - level <= max_difference for level, next_level in zip(levels, levels[1:]))

def is_level_decreasing(levels: list[int], max_difference: int) -> bool:
   return all(level > next_level and level - next_level <= max_difference for level, next_level in zip(levels, levels[1:]))

if __name__ == "__main__":
    main()